syntax = "proto3";

package csv_processor;

service CsvProcessor {
  rpc ProcessCsv (stream CsvChunk) returns (ProcessCsvResponse);
  rpc GetProcessingResult (GetProcessingResultRequest) returns (GetProcessingResultResponse);
  rpc DownloadResult (DownloadResultRequest) returns (stream CsvChunk);
}

message CsvChunk {
  bytes data = 1;
}

message ProcessCsvResponse {
  string task_id = 1;
  string status = 2;
}

message GetProcessingResultRequest {
  string task_id = 1;
}

message Progress {
  int32 lines_processed = 1;
  int32 departments = 2;
  float time_elapsed = 3;
}

message GetProcessingResultResponse {
  string processed_csv_path = 1;
  bool completed = 2;
  string status = 3;
  Progress progress = 4;
}

message DownloadResultRequest {
  string task_id = 1;
}